### PR TITLE
pgdb: cover transactions with `inTransaction` function to simplify error handling

### PR DESCRIPTION
**Pull request by rumyantseva**
# Description

**Note:** This PR might be hard to review, as a lot of lines of the `pgdb` package were changed. Let me know how I can help you with reviewing this PR.

There is no special issue for this fix, but it's done as a part of ref #372.

## inTransaction

As now we use transactions in many places, handling potential errors (when a transaction begins, commits or rollbacks) takes quite a lot of lines of code (:D) and requires some accuracy.

To simplify this approach, this PR offers an `inTransaction` function to "wrap" any function with a transaction.

## Querier

`pgxtype.Querier` is introduced instead of `pgx.Tx` to unify function signatures.

## Dealing with errors

As some of the `pgdb` errors might be wrapped, the comparison of errors was refactored to `errors.Is` and `errors.As` instead of `==`.

## Not in scope

This PR only offers a refactoring for the existing transactions, adding new transactions is not in the scope of this PR.

## Readiness checklist

* [x] I set assignee, reviewers, labels, project and sprint.
* [x] I added tests for new functionality or bugfixes.
* [x] I ran `task all`, and it passed.
* [x] I added/updated documentation for exported and unexported functions, variables, types, etc.
* [x] I checked complex documentation rendering with `bin/godoc -index -play -timestamps -http=127.0.0.1:6060`.

### CODE REVIEW

**Comment by rumyantseva** [comment]
@AlekSi @w84thesun please take a look at this quick fix. Let's not log errors if there are no errors (otherwise, the logs look quite confusing).

File: internal/handlers/pg/pgdb/pool.go [hunk 1]
```diff
@@ -291,10 +291,14 @@ func (pgPool *Pool) CreateDatabase(ctx context.Context, db string) error {
...
 		if err != nil {
-			pgPool.logger.Error("failed to perform rollback", zap.Error(tx.Rollback(ctx)))
+			if rerr := tx.Rollback(ctx); rerr != nil {
+				pgPool.logger.Error("failed to perform rollback", zap.Error(rerr))
+			}
 			return
 		}
-		pgPool.logger.Error("failed to perform commit", zap.Error(tx.Commit(ctx)))
+		if cerr := tx.Commit(ctx); cerr != nil {
+			pgPool.logger.Error("failed to perform commit", zap.Error(cerr))
```
Comment (AlekSi) [inline internal/handlers/pg/pgdb/pool.go:300]: I think `CreateDatabase` should error in that case. That would require a named result, etc.

And I think we should call `Rollback` there too

Comment (rumyantseva) [reply internal/handlers/pg/pgdb/pool.go:300]: Yeah, good point...

**Review by w84thesun — COMMENTED** [review]
I think `inTransaction` method makes it difficult to read the code, as you need to keep it in mind when you read other methods code. Maybe introducing function like `finishTx(tx pgx.Tx, err error, method string)` would be enough?

**Comment by AlekSi** [comment]
#844 was merged, please resolve merge conflicts and replace a few more `==` with `errors.Is`

File: internal/handlers/pg/msg_create.go [hunk 2]
```diff
@@ -68,16 +69,17 @@ func (h *Handler) MsgCreate(ctx context.Context, msg *wire.OpMsg) (*wire.OpMsg,
...
 	}
 
-	if err := h.pgPool.CreateDatabase(ctx, db); err != nil && err != pgdb.ErrAlreadyExist {
+	if err := h.pgPool.CreateDatabase(ctx, db); err != nil && !errors.Is(err, pgdb.ErrAlreadyExist) {
 		return nil, lazyerrors.Error(err)
 	}
 
-	if err = h.pgPool.CreateCollection(ctx, db, collection); err != nil {
-		if err == pgdb.ErrAlreadyExist {
+	// TODO: use a transaction, not just h.pgPool
```
Comment (w84thesun) [inline internal/handlers/pg/msg_create.go:76]: Could you please add an issue here?

File: internal/handlers/pg/pgdb/pool.go [hunk 3]
```diff
@@ -692,54 +663,43 @@ func (pgPool *Pool) InsertDocument(ctx context.Context, db, collection string, d
...
 
 	if !exists {
-		if err := pgPool.CreateDatabase(ctx, db); err != nil && err != ErrAlreadyExist {
+		if err := pgPool.CreateDatabase(ctx, db); err != nil && !errors.Is(err, ErrAlreadyExist) {
 			return lazyerrors.Error(err)
 		}
 
-		if err := pgPool.CreateCollection(ctx, db, collection); err != nil {
-			if err == ErrAlreadyExist {
+		// TODO: use a transaction instead of pgPool.
```
Comment (w84thesun) [inline internal/handlers/pg/pgdb/pool.go:670]: And here too.

File: internal/handlers/pg/pgdb/pool.go [hunk 4]
```diff
@@ -498,12 +489,13 @@ func (pgPool *Pool) CreateTableIfNotExist(ctx context.Context, db, collection st
...
 	// but keep in mind that it can be created in concurrent connection.
 
-	if err := pgPool.CreateDatabase(ctx, db); err != nil && err != ErrAlreadyExist {
+	if err := pgPool.CreateDatabase(ctx, db); err != nil && !errors.Is(err, ErrAlreadyExist) {
 		return false, lazyerrors.Error(err)
 	}
 
-	if err := pgPool.CreateCollection(ctx, db, collection); err != nil {
-		if err == ErrAlreadyExist {
+	// TODO: use a transaction instead of pgPool.
```
Comment (w84thesun) [inline internal/handlers/pg/pgdb/pool.go:496]: Please add an issue link here.

Comment (AlekSi) [reply internal/handlers/pg/pgdb/pool.go:496]: I will create an issue

Comment (AlekSi) [reply internal/handlers/pg/pgdb/pool.go:496]: @rumyantseva #866

**Review by w84thesun — APPROVED** [review]
LGTM.

**Review by seeforschauer — APPROVED** [review]

**Review by AlekSi — APPROVED** [review]

**Pull request merged**